src/main/java/net/minecraftforge/common/BiomeDictionary.java

```diff
@@ -21,132 +21,133 @@
+
+        public static final Type OCEAN = new Type("OCEAN");
+        public static final Type RIVER = new Type("RIVER");
+        /**
+         * A general tag for all water-based biomes. Shown as present if OCEAN or RIVER are.
+         **/
+        public static final Type WATER = new Type("WATER", OCEAN, RIVER);
 
         /*Generic types which a biome can be*/
-        MESA,
-        FOREST,
-        PLAINS,
-        MOUNTAIN,
-        HILLS,
-        SWAMP,
-        SANDY,
-        SNOWY,
-        WASTELAND,
-        BEACH;
-
-        private List<Type> subTags;
-
-        private Type(Type... subTags)
+        public static final Type MESA = new Type("MESA");
+        public static final Type FOREST = new Type("FOREST");
+        public static final Type PLAINS = new Type("PLAINS");
+        public static final Type MOUNTAIN = new Type("MOUNTAIN");
+        public static final Type HILLS = new Type("HILLS");
+        public static final Type SWAMP = new Type("SWAMP");
+        public static final Type SANDY = new Type("SANDY");
+        public static final Type SNOWY = new Type("SNOWY");
+        public static final Type WASTELAND = new Type("WASTELAND");
+        public static final Type BEACH = new Type("BEACH");
+
+        private static final Map<String, Type> byName = new HashMap<String, Type>();
+
+        private final String name;
+        private final List<Type> subTags;
+        final Set<Biome> biomes = new HashSet<Biome>();
+        final Set<Biome> biomesUn = Collections.unmodifiableSet(biomes);
+
+        private Type(String name, Type... subTags)
         {
-            this.subTags = Arrays.asList(subTags);
+            this.name = name;
+            this.subTags = ImmutableList.copyOf(subTags);
+
+            byName.put(name, this);
         }
 
         private boolean hasSubTags()
         {
-            return subTags != null && !subTags.isEmpty();
+            return !subTags.isEmpty();
+        }
+
+        public String name()
+        {
+            return name;
         }
 
         /**
          * Retrieves a Type value by name,
          * if one does not exist already it creates one.
          * This can be used as intermediate measure for modders to
          * add their own category of Biome.
-         *
+         * <p>
          * There are NO naming conventions besides:
-         *   MUST be all upper case (enforced by name.toUpper())
-         *   NO Special characters. {Unenforced, just don't be a pain, if it becomes a issue I WILL
-         *                             make this RTE with no worry about backwards compatibility}
-         *
+         * MUST be all upper case (enforced by name.toUpper())
+         * NO Special characters. {Unenforced, just don't be a pain, if it becomes a issue I WILL
+         * make this RTE with no worry about backwards compatibility}
+         * <p>
          * Note: For performance sake, the return value of this function SHOULD be cached.
```

Comment:
Probably not as much anymore, since it's just a hash lookup now. But if you are calling this a lot during worldgen it might be better to put it in a static final field, if only for cleaner code.